Use the new autoconf macros

dridi opened the pull request:
Split in individual commits, details (if any) can be found in commit messages for each change. These changes technically also belong in the 4.1 branch since 4.1.4-beta1 includes the new macros too. The question is, do we support older 4.1 releases? If yes we should keep the old build system in the 4.1 branch.

> 7 files changed, 63 insertions(+), 131 deletions(-)

Summoning: @scn @gquintard @aondio

dridi committed:
Update autoconf initialization

Require a more recent autoconf, and name the dist archive vmod-example
to match the usual package naming scheme. Update the copyright notice.

dridi committed:
Update obsolete macro

dridi committed:
Move some generated files out of the root dir

dridi committed:
Remove AC_CANONICAL_SYSTEM

It is now replaced by AC_CANONICAL_TARGET but regardless, the $target
variable isn't used anywhere.

dridi committed:
Require a relatively recent version of automake

Make it choke on warnings, set up the test driver and use silent rules.

dridi committed:
Require a relatively recent version of libtool

Libtool itself requires most of the compilation toolchain, so we don't
need to duplicate effort.

dridi committed:
PKG_CHECK_VAR is now provided by varnish-legacy.m4

dridi committed:
Remove check for unused header

dridi committed:
Get a proper LD_LIBRARY_PATH to run the test suite

dridi committed:
Use the new autoconf macros

dridi committed:
Take a shortcut

dridi committed:
Override RST2MAN using a --with argument

dridi committed:
Update the README to reflect today's changes

dridi committed:
The README is no longer a man page

aondio commented:
I've played around a bit with this PR and it looks good to me.

Little request: travis is failing, can you please update the travis.yml which is still trying to build the vmod against 4.1?

dridi commented:
@aondio good catch, but I'm afraid it will have to wait for packages before this can be merged:

https://repo.varnish-cache.org/ubuntu/dists/precise/

dridi commented:
I forgot to summon interested parties: @fgsch @nigoroll

nigoroll commented:
Oh damn, I'm so sorry, @Dridi, I had forgot about your PR and just did the thing, and now I'm fighting with travis.
BIG APOLOGIES AGAIN!

nigoroll commented:
ok, due to the exeptional fuckup from my side I've force-pushed back to dc4f97b6b2b93c58fe72fddd50b64e6f51180f68
Will now take the opportunity to use my independent changes for a review of your work, @Dridi

nigoroll commented:
Re Travis: The last I had was this in `.travis.yml`:

```
diff --git a/.travis.yml b/.travis.yml
index c0be8b7..18b0556 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -2,15 +2,13 @@ language: c

 before_install:
   - sudo apt-get update -q
-  - sudo apt-get install -qq apt-transport-https python-docutils
-  - curl https://repo.varnish-cache.org/debian/GPG-key.txt | sudo apt-key add -
-  - echo "deb https://repo.varnish-cache.org/ubuntu/ precise varnish-4.1" | sudo tee /etc/apt/sources.list.d/varnish-cache.list
-  - sudo apt-get -q update
-  - sudo apt-get install varnish libvarnishapi-dev
+  - sudo apt-get install -qq apt-transport-https python-docutils libjemalloc1
+  - wget -q https://repo.varnish-cache.org/pkg/5.0.0/varnish_5.0.0-1_amd64.deb https://repo.varnish-cache.org/pkg/5.0.0/varnish-dev_5.0.0-1_amd64.deb
+  - sudo dpkg -i varnish_5.0.0-1_amd64.deb varnish-dev_5.0.0-1_amd64.deb


 before_script:
-  - ./autogen.sh
+  - ./bootstrap --prefix=/usr
   - ./configure --prefix=/usr
   - make -j4

```

but still got:

```
Error:
Message from VCC-compiler:
/usr/sbin/varnishd: symbol lookup error: /usr/lib/varnish/libvarnish.so: undefined symbol: pcre_free_study
```

This sounds like the we need a separate build for precise

nigoroll commented:
- would suggest to switch to bootstrap and improve the script
  https://github.com/nigoroll/libvmod-example/blob/my_5.0/autogen.sh
  https://github.com/nigoroll/libvmod-example/blob/my_5.0/bootstrap
- shouldnt we keep
  `m4_ifndef([VARNISH_VMOD_INCLUDES], AC_MSG_ERROR([Need varnish.m4 -- see README.rst]))`
  in configure.ac?
- you've been a bit to eager with `.gitignore`
  Other than that, your banch WFM and LGTM - probably much better than mine :)

dridi committed:
Make the master branch 5.0-only

Although it may build with 4.1.2+, stick to the branching scheme.

dridi commented:
Regarding Travis, I have a [module](https://github.com/Dridi/libvmod-querystring/blob/master/.travis.yml) using the new macros that builds against 5.0.0 so depending on preferences I can install it from a package too.

> This sounds like the we need a separate build for precise

I'm not sure which direction we are taking with packaging, but it seems that we mainly ship dist archives and raw deb/rpm packages (not in apt/yum repositories). And debs don't seem to embed the target in their file names.

See https://repo.varnish-cache.org/pkg/5.0.0/

Also I think the new macros don't work on precise, `automake` is too old (unless of course you install a more recent version). But it builds fine from a dist tarball.

> would suggest to switch to bootstrap and improve the script

The same module has a simpler [bootstrap](https://github.com/Dridi/libvmod-querystring/blob/master/bootstrap) script that doesn't check the `automake` version because it'll be done at configure time. Also it contains no bashisms that I know of (and just in case, `$(...)` is not a bashism).

Also, I don't see the point of keeping `autogen.sh` if we update the README accordingly.

> shouldnt we keep [...] in configure.ac?

No. The old and new macros don't mix well together, but both work fine on their own. I didn't remove it, I simply kept `AC`, `AM` and `LT` macros together before dealing with Varnish.

See [this check](https://github.com/varnishcache/libvmod-example/pull/2/commits/030bd14bce301796b40106beb25a2c2c9f52eb66#diff-67e997bcfdac55191033d57a16d1408aR29) before the use of `VARNISH` macros.

> you've been a bit to eager with `.gitignore`

Are you sure? I thought I had only removed irrelevant patterns. Don't you have leftovers from a previous branch instead?

Thanks for the review!

nigoroll commented:
@Dridi, I have re-checked:
- you were right regarding `.gitignore` :) thx
- I _suspect_ we will receive complaints from users of old autocrap tools, but personally I am fine with modernizing and documenting the requirements
- adding a varnish build to travis seems like a sensible thing to do while we don't have the right binary build available
  So I'd appreciate a merge sooner than later in order to use this as a basis for other vmods.
  Thank you!

dridi committed:
Install Varnish from source in Travis CI

The Varnish version appears only once, making it hard to break on
Varnish upgrades.

dridi commented:
I made this PR merge-able after @nigoroll's last comment.

nigoroll commented on bootstrap:
```diff
@@ -0,0 +1,25 @@
+	echo "libtoolize: command not found, falling back to glibtoolize" >&2
+	alias libtoolize=glibtoolize
+fi
+
```
i think we should have `set -x` here, otherwise it will not be clear where the script fails, if it does

dridi replied:
Done in a397c91.

nigoroll commented on autogen.sh:
```diff
@@ -1,53 +0,0 @@
-fi
-
-# check for varnishapi.m4 in custom paths
-dataroot=$(pkg-config --variable=datarootdir varnishapi 2>/dev/null)
```
I think we need this...

dridi replied:
We don't, pkg-config will be used by the `configure` script.

nigoroll commented on autogen.sh:
```diff
@@ -1,53 +0,0 @@
-	exit 1
-fi
-set -ex
-aclocal -I m4 -I ${dataroot}/aclocal
```
... and this, otherwise I don't see how a custom varnish install location would work

```
slink@haggis:~/Devel/varnish-git/libvmod-example (dridi_varnish-5)$ export PKG_CONFIG_PATH=$PREFIX/lib/pkgconfig
slink@haggis:~/Devel/varnish-git/libvmod-example (dridi_varnish-5)$ ./bootstrap 
libtoolize: putting auxiliary files in AC_CONFIG_AUX_DIR, `build-aux'.
libtoolize: copying file `build-aux/ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIR, `m4'.
libtoolize: copying file `m4/libtool.m4'
libtoolize: copying file `m4/ltoptions.m4'
libtoolize: copying file `m4/ltsugar.m4'
libtoolize: copying file `m4/ltversion.m4'
libtoolize: copying file `m4/lt~obsolete.m4'
src/Makefile.am:4: error: 'vmod_LTLIBRARIES' is used but 'vmoddir' is undefined
automake: warnings are treated as errors
src/Makefile.am:6: warning: variable 'libvmod_example_la_SOURCES' is defined but no program or
src/Makefile.am:6: library has 'libvmod_example_la' as canonical name (possible typo)
src/Makefile.am:9: warning: variable 'nodist_libvmod_example_la_SOURCES' is defined but no program or
src/Makefile.am:9: library has 'libvmod_example_la' as canonical name (possible typo)

```

dridi replied:
You didn't set the `ACLOCAL_PATH`, so you are picking `varnish.m4` from your `/usr` installation of Varnish. It's documented in the README.

nigoroll replied:
yes, you are right, I missed this.
But I still think we should keep the original code to automatically add the aclocal include.
The error message is obscure and being more admin-friendly should be a good reason for a couple of lines more in a shell script.

dridi replied:
> yes, you are right, I missed this.

That makes the review more interesting :)

> But I still think we should keep the original code to automatically add the aclocal include.

I get the user experience point, but in this case that would actually add the `aclocal` include _manually_. Because it's otherwise picked up _automatically_ from the environment. And said environment could contain `*_PATH`s with more than one entry (let's imagine another dependency than Varnish not being installed in `/usr`...)

The one place where it is done is `Makefile.am`, because once your build tree is configured you may `make` without the `*_PATH`s set in your environment:

```
ACLOCAL_AMFLAGS = -I m4 -I ${VARNISHAPI_DATAROOTDIR}/aclocal
```

> The error message is obscure and being more admin-friendly

In this case I'd rather update the README on common problems.

nigoroll replied:
... if you insist ...
I think we're making things more complicated than they need to be and I am not happy about it personally.
But I want to get ahead, so just do the merge.

dridi replied:
I don't insist, I had originally left `autogen.sh` untouched, I will remove those changes and keep it strictly to using the new `autoconf` macros.

Merging this afternoon unless someone objects now that it has been approved by @nigoroll.
